File: app/services/move_finder.rb
class MoveFinder
  MOVE_SET = [ [ 1, 2 ], [ 2, 1 ], [ 2, -1 ], [ 1, -2 ], [ -1, -2 ], [ -2, -1 ], [ -2, 1 ], [ -1, 2 ] ].freeze

  attr_accessor :x, :y

  def initialize(square:)
    @x = square.x
    @y = square.y
  end

  def legal_moves
    move_candidates
      .select { |x, y| (1..8).cover?(x) && (1..8).cover?(y) }
      .map { |x, y| Square.new(x:, y:) }
  end

  def move_candidates
    MOVE_SET.map { |move| move(move) }
  end

  def move(move)
    [ x + move[0], y + move[1] ]
  end
end

File: app/javascript/game/move_finder.js
import { Square } from "#game/square"

const MOVE_SET = [ [ 1, 2 ], [ 2, 1 ], [ 2, -1 ], [ 1, -2 ], [ -1, -2 ], [ -2, -1 ], [ -2, 1 ], [ -1, 2 ] ]

export class MoveFinder {
  constructor(square) { this.square = square }

  legalMoves() {
    return this.moveCandidates()
      .filter(([ x, y ]) => x >= 1 && x <= 8 && y >= 1 && y <= 8)
      .map(([ x, y ]) => new Square(x, y))
  }

  moveCandidates() {
    return MOVE_SET.map(([ dx, dy ]) => [ this.square.x + dx, this.square.y + dy ])
  }
}
